### PR TITLE
Interval implementation

### PR DESCRIPTION
**Pull request by mkleen**
## Checklist

 - [ ] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [ ] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: sql/src/main/java/io/crate/protocols/postgres/types/IntervalType.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    byte[] encodeAsUTF8Text(@Nonnull Object value) {
+        StringBuffer sb = new StringBuffer();
+        PERIOD_FORMATTER.printTo(sb, (ReadablePeriod) value);
+        return sb.toString().getBytes(StandardCharsets.UTF_8);
```
Comment (mkleen) [inline sql/src/main/java/io/crate/protocols/postgres/types/IntervalType.java]: This is just a temporary solution, i still need to port over the psql implementation.

File: common/src/main/java/io/crate/interval/IntervalParser.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+ * agreement.
+ */
+
+package io.crate.interval;
+
+import org.joda.time.Period;
+
+import java.math.BigDecimal;
+
+public class IntervalParser {
```
Comment (mkleen) [inline common/src/main/java/io/crate/interval/IntervalParser.java]: This seems a bit odd in the beginning since there are no interfaces and no inheritance with the other parser, but i preferred to keep as a set of stateless functions. You can see the parser more like packages grouped together. If you don't like it I can convert it to a normal java-style inheritance.

File: common/src/main/java/io/crate/interval/IntervalParser.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        if (start == Precision.YEAR && end == Precision.MONTH) {
+            return Period.years(period.getYears()).withMonths(period.getMonths());
+        }
+        if (start == Precision.MONTH && end == null) {
+            return Period.years(period.getYears()).withMonths(period.getMonths());
+        }
+        if (start == Precision.DAY && end == null) {
+            return new Period(period).withHours(0).withMinutes(0).withSeconds(0).withMillis(0);
+        }
+        if (start == Precision.DAY && end == Precision.HOUR) {
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/interval/IntervalParser.java]: How about having a single `if (start == Precision.DAY)` and then within that branch handle the different `end` cases to avoid repeated `start == DAY` checks.

File: common/src/main/java/io/crate/interval/IntervalParser.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+            return new Period(period).withMinutes(0).withSeconds(0).withMillis(0);
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/interval/IntervalParser.java]: I think the `new Period(period)` calls are not necessary. `.withXY(..)` already creates a copy and doesn't mutate instances.

File: common/src/main/java/io/crate/interval/IntervalParser.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        if (start == Precision.MINUTE && end == Precision.SECOND) {
+            return new Period(period).withMillis(0);
+        }
+        if (start == Precision.SECOND && end == null) {
+            return new Period(period).withMillis(0);
+        }
+        throw new IllegalArgumentException(String.format("Invalid start and end combination", start, end));
+    }
+
+    static int parseMilliSeconds(String value) throws NumberFormatException {
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/interval/IntervalParser.java]: I think this could use some doc to explain the format `value` is supposed to be in.

File: common/src/main/java/io/crate/interval/IntervalParser.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        BigDecimal decimal = new BigDecimal(value);
+        return decimal
+            .subtract(new BigDecimal(decimal.intValue()))
+            .multiply(new BigDecimal(1000)).intValue();
+    }
+
+    static int parseInteger(String value) {
+        return new BigDecimal(value).intValue();
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/interval/IntervalParser.java]: Same here. To me it is unclear why we'd go through `BigDecimal` instead of parsing the int value directly.

File: common/src/main/java/io/crate/interval/IsoIntervalParser.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import org.joda.time.Period;
+import org.joda.time.format.ISOPeriodFormat;
+
+import static io.crate.interval.IntervalParser.roundToPrecision;
+
+class IsoIntervalParser {
+
+    private IsoIntervalParser() { }
+
+    static Period apply(String value) {
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/interval/IsoIntervalParser.java]: Not sure if this indirection is worth it, why not call `standard().parsePeriod(value)` directly on the call-sites?

File: common/src/main/java/io/crate/interval/IsoIntervalParser.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        return ISOPeriodFormat.standard().parsePeriod(value);
+    }
+
+    static Period apply(String value, IntervalParser.Precision start, IntervalParser.Precision end) {
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/interval/IsoIntervalParser.java]: Same here.

File: common/src/main/java/io/crate/interval/IntervalParser.java [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        SECOND
+    }
+
+    private IntervalParser() { }
+
+    public static Period apply(String value) {
+        return apply(value, null, null);
+    }
+
+    public static Period apply(String value, Precision start, Precision end) {
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/interval/IntervalParser.java]: please add `@Nullable` annotations for args that can be null

File: common/src/main/java/io/crate/interval/NumericalIntervalParser.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        if (start == IntervalParser.Precision.YEAR && end == IntervalParser.Precision.MONTH) {
+            return Period.months(value);
+        }
+        if (start == IntervalParser.Precision.MONTH && end == null) {
+            return Period.months(value);
+        }
+        if (start == IntervalParser.Precision.DAY && end == null) {
+            return Period.days(value);
+        }
+        if (start == IntervalParser.Precision.DAY && end == IntervalParser.Precision.HOUR) {
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/interval/NumericalIntervalParser.java]: See comment in the other class about doing the `if (start == ...)` check only once.

File: common/src/main/java/io/crate/interval/NumericalIntervalParser.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        if (start == IntervalParser.Precision.MINUTE && end == IntervalParser.Precision.SECOND) {
+            return Period.seconds(value);
+        }
+        if (start == IntervalParser.Precision.SECOND && end == null) {
+            return roundToSecondsWithMillis(value, millis);
+        }
+        throw new IllegalArgumentException(String.format("Invalid start %s and end %s combination", start, end));
+    }
+
+    private static Period roundToSecondsWithMillis(int seconds, int millis) {
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/interval/NumericalIntervalParser.java]: the `round` here seems a bit confusing. Isn't this keeping the full precision of both seconds & millis?

File: common/src/main/java/io/crate/interval/PGIntervalParser.java [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+import static io.crate.interval.IntervalParser.parseMilliSeconds;
+import static io.crate.interval.IntervalParser.roundToPrecision;
+
+class PGIntervalParser {
+
+    static Period apply(String value, IntervalParser.Precision start, IntervalParser.Precision end) {
+        return roundToPrecision(apply(value), start, end);
+    }
+
+    static Period apply(String value) {
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/interval/PGIntervalParser.java]: I think the methods here could also use some docs to explain the format of the strings they're parsing.

File: common/src/main/java/io/crate/interval/PGIntervalParser.java [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+                        continue;
+                    }
+                    // This handles hours, minutes, seconds and microseconds for
+                    // ISO intervals
+                    int offset = (token.charAt(0) == '-') ? 1 : 0;
+
+                    hours = nullSafeIntGet(token.substring(offset + 0, endHours));
+                    minutes = nullSafeIntGet(token.substring(endHours + 1, endHours + 3));
+
+                    // Pre 7.4 servers do not put second information into the results
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/interval/PGIntervalParser.java]: Is this relevant for us?

File: common/src/main/java/io/crate/types/IntervalType.java [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+            return null;
+        }
+    }
+
+    @Override
+    public void writeValueTo(StreamOutput out, Period p) throws IOException {
+        if (p == null) {
+            out.writeBoolean(false);
+        } else {
+            out.writeLong(p.getSeconds());
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/types/IntervalType.java]: I think this needs to be extended to properly stream the full object. None of `getSeconds`, `getDays` or `getMonths` are a total. So if a `Period` for example contains years=7, those 7 years would be lost.

File: common/src/main/java/io/crate/types/IntervalType.java [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+            out.writeInt(p.getDays());
+            out.writeInt(p.getMonths());
+        }
+    }
+
+    @Override
+    public int fixedSize() {
+        return 32;
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/types/IntervalType.java]: Is this still accurate?

File: common/src/test/java/io/crate/interval/PGIntervalParserTest.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+
+public class PGIntervalParserTest {
+
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
+
+    @Test
+    public void test_psql_format_from_string() {
+        Period period = PGIntervalParser.apply("@ 1 year 1 mon 1 day 1 hour 1 minute 1 secs");
+        Assert.assertEquals(period,
```
Comment (mfussenegger) [inline common/src/test/java/io/crate/interval/PGIntervalParserTest.java]: We generally prefer `assertThat(..., is(..))` for better error messages.

File: sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java [hunk 17]
```diff
@@ -1054,12 +1116,12 @@ public Node visitArraySubquery(SqlBaseParser.ArraySubqueryContext ctx) {
...
     /*
-    * case sensitivity like it is in postgres
-    * see also http://www.thenextage.com/wordpress/postgresql-case-sensitivity-part-1-the-ddl/
-    *
-    * unfortunately this has to be done in the parser because afterwards the
-    * knowledge of the IDENT / QUOTED_IDENT difference is lost
-    */
+     * case sensitivity like it is in postgres
+     * see also http://www.thenextage.com/wordpress/postgresql-case-sensitivity-part-1-the-ddl/
+     *
```
Comment (mfussenegger) [inline sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: Please avoid unrelated formatting changes. (Especially on a PR against a feature branch).

Maybe you can do those separate in a PR against master?

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 18]
```diff
@@ -919,6 +927,34 @@ public Symbol visitObjectLiteral(ObjectLiteral node, ExpressionAnalysisContext c
                 context);
         }
 
+        private final Map<IntervalLiteral.IntervalField, IntervalParser.Precision> INTERVAL_FIELDS =
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]:
```suggestion
        private static final Map<IntervalLiteral.IntervalField, IntervalParser.Precision> INTERVAL_FIELDS =
```

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 19]
```diff
@@ -919,6 +927,34 @@ public Symbol visitObjectLiteral(ObjectLiteral node, ExpressionAnalysisContext c
...
+                DAY, IntervalParser.Precision.DAY,
+                HOUR, IntervalParser.Precision.HOUR,
+                MINUTE, IntervalParser.Precision.MINUTE,
+                SECOND, IntervalParser.Precision.SECOND
+            );
+
+        @Override
+        public Symbol visitIntervalLiteral(IntervalLiteral node, ExpressionAnalysisContext context) {
+            String value = node.getValue();
+            if (value == null || value.isEmpty() || value.isBlank()) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: Isn't the IntervalParser already taking care of  empty or blank cases?

File: sql/src/main/java/io/crate/protocols/postgres/types/IntervalType.java [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+    @Override
+    public int typArray() {
+        return PGArray.INTERVAL_ARRAY.oid();
+    }
+
+    @Override
+    public int writeAsBinary(ByteBuf buffer, @Nonnull Object value) {
+        Period period = (Period) value;
+        buffer.writeInt(TYPE_LEN);
+        buffer.writeLong(period.getSeconds());
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/protocols/postgres/types/IntervalType.java]: See comment in the `IntervalType` about streaming. I think we need to build total values here in order to not loose information.

File: sql/src/test/java/io/crate/testing/DataTypeTesting.java [hunk 21]
```diff
@@ -128,6 +131,11 @@
                     map.put("x", innerValueGenerator.get());
                     return (T) map;
                 };
+            case IntervalType.ID:
+                return () -> {
+                    return (T) new Period().withMillis(1000);
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/testing/DataTypeTesting.java]: Maybe we can randomize this a bit?

File: common/src/main/java/io/crate/interval/IntervalParser.java (hunk 2 above)
Comment (mfussenegger) [reply common/src/main/java/io/crate/interval/IntervalParser.java]: I like it this way with stateless functions 👍

File: sql/src/main/java/io/crate/protocols/postgres/types/IntervalType.java (hunk 20 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/protocols/postgres/types/IntervalType.java]: Maybe extend `testByteReadWrite`  in `PGTypesTest` as part of this, so that the round-trip of encode/decode is tested.

File: sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java [hunk 22]
```diff
@@ -223,6 +224,64 @@ public Node visitBegin(SqlBaseParser.BeginContext context) {
...
+            case SqlBaseLexer.MINUS:
+                return IntervalLiteral.Sign.MINUS;
+            case SqlBaseLexer.PLUS:
+                return IntervalLiteral.Sign.PLUS;
+            default:
+                throw new IllegalArgumentException("Unsupported sign: " + token.getText());
+        }
+    }
+
+    private static IntervalLiteral.IntervalField getIntervalFieldType(Token token) {
```
Comment (seut) [inline sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: Given that the `IntervalField` is never used but directly converted to a `IntervalParser.Precision` at the expression analyzer, maybe it makes sense to use `IntervalParser.Precision.` already here and get rid of the `IntervalField`?

Comment (mkleen) [reply sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: Yes, good idea.

File: common/src/main/java/io/crate/interval/IntervalParser.java (hunk 2 above)
Comment (mkleen) [reply common/src/main/java/io/crate/interval/IntervalParser.java]: final - good idea.

File: common/src/main/java/io/crate/interval/IsoIntervalParser.java [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+package io.crate.interval;
+
+import org.joda.time.Period;
+import org.joda.time.format.ISOPeriodFormat;
+
+import static io.crate.interval.IntervalParser.roundToPrecision;
+
+class IsoIntervalParser {
```
[the remark this answers is not in the document]
Comment (mkleen) [reply common/src/main/java/io/crate/interval/IsoIntervalParser.java]: yes, good point.

File: common/src/test/java/io/crate/interval/PGIntervalParserTest.java [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+
+public class PGIntervalParserTest {
+
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
```
[the remark this answers is not in the document]
Comment (mkleen) [reply common/src/test/java/io/crate/interval/PGIntervalParserTest.java]: Good catch, will remove it.

File: sql-parser/src/main/java/io/crate/sql/tree/IntervalLiteral.java [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+
+    public Sign getSign() {
+        return sign;
+    }
+
+    public IntervalField getStartField() {
+        return startField;
+    }
+
+    public @Nullable IntervalField getEndField() {
```
[the remark this answers is not in the document]
Comment (mkleen) [reply sql-parser/src/main/java/io/crate/sql/tree/IntervalLiteral.java]: 👍

File: sql-parser/src/main/java/io/crate/sql/tree/IntervalLiteral.java [hunk 26]
```diff
@@ -0,0 +1,121 @@
...
+        DAY,
+        HOUR,
+        MINUTE,
+        SECOND
+    }
+
+    private final String value;
+    private final Sign sign;
+    private final IntervalField startField;
+    private final IntervalField endField;
```
[the remark this answers is not in the document]
Comment (mkleen) [reply sql-parser/src/main/java/io/crate/sql/tree/IntervalLiteral.java]: 👍

File: sql-parser/src/test/java/io/crate/sql/IntervalLiteralTest.java [hunk 27]
```diff
@@ -0,0 +1,135 @@
...
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import java.util.Optional;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.nullValue;
+import static org.hamcrest.core.Is.is;
+
+public class IntervalLiteralTest {
```
[the remark this answers is not in the document]
Comment (mkleen) [reply sql-parser/src/test/java/io/crate/sql/IntervalLiteralTest.java]: 👍

File: common/src/main/java/io/crate/interval/IntervalParser.java (hunk 3 above)
Comment (mfussenegger) [reply common/src/main/java/io/crate/interval/IntervalParser.java]: Yep it won't be shorter, but it safes some CPU cycles and personally I think it is also easier to read, as you don't have to mentally repeatedly parse 2 expressions.

File: sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java (hunk 22 above)
Comment (mkleen) [reply sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: This would introduce a new dependency between sql-parser and common which is currently not there. If this ok with you, i will do it.

Comment (seut) [reply sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: Hm right, so ok lets leave it like that.

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java (hunk 18 above)
Comment (mkleen) [reply sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: Inner classes cannot have static members, but maybe I remove `IntervalLiteral.IntervalField` completely and use the `IntervalParser.Precision` everywhere, then the whole map can go away.

File: common/src/main/java/io/crate/types/IntervalType.java (hunk 15 above)
Comment (mkleen) [reply common/src/main/java/io/crate/types/IntervalType.java]: This is estimated with JOL plugin based `joda.Period`

File: sql-parser/src/test/java/io/crate/sql/parser/TestStatementBuilder.java [hunk 28]
```diff
@@ -856,6 +856,11 @@ public void testSelectStmtBuilder() {
         printStatement("select * from t where ? !~ ?");
     }
 
+    @Test
+    public void testIntervalLiteral() {
+        printStatement("select interval '1' HOUR");
```
[the remark this answers is not in the document]
Comment (mkleen) [reply sql-parser/src/test/java/io/crate/sql/parser/TestStatementBuilder.java]: 👍

File: sql-parser/build.gradle [hunk 29]
```diff
@@ -20,6 +20,7 @@ dependencies {
     compile "com.google.code.findbugs:jsr305:${versions.jsr305}"
     compile "com.google.guava:guava:${versions.guava}"
 
+    testImplementation project(':integration-testing')
```
Comment (mfussenegger) [inline sql-parser/build.gradle]: I'm not sure if it is worth pulling in the dependency to safe us from having to declare the `expectedException`. `:integration-testing` is very heavy and pulls in a lot.

I've had in mind to build some tooling that re-uses the sql-parser package standalone. I never got around to doing any of it, but I think it would be cool if we could keep this fairly isolated in terms of dependencies.

File: sql/src/test/java/io/crate/lucene/ArrayLengthQueryTest.java [hunk 30]
```diff
@@ -273,6 +274,10 @@ public void testArrayLengthEq0ReturnsNoElements() throws Exception {
     @Test
     public void testArrayLengthWithAllSupportedTypes() throws Exception {
         for (DataType<?> type : DataTypeTesting.ALL_TYPES_EXCEPT_ARRAYS) {
+            // This is temporary as long as interval is not fully implemented
+            if(type.equals(DataTypes.INTERVAL)) {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/lucene/ArrayLengthQueryTest.java]: Could also use `DataTypes.STORAGE_UNSUPPORTED` here, once we support interval storage the test would then reflect the change automatically (otherwise we'll forget about this "temporary" skipping)

File: sql-parser/src/test/java/io/crate/sql/parser/TestStatementBuilder.java (hunk 28 above)
[the remark this answers is not in the document]
Comment (mkleen) [reply sql-parser/src/test/java/io/crate/sql/parser/TestStatementBuilder.java]: There is extensive testing in `IntervalAnalysisTest` and the interval parser tests.

File: common/src/main/java/io/crate/interval/PGIntervalParser.java [hunk 31]
```diff
@@ -99,6 +99,8 @@ static Period apply(String value) {
                         months = nullSafeIntGet(valueToken);
                     } else if (token.startsWith("day")) {
                         days = nullSafeIntGet(valueToken);
+                    } else if (token.startsWith("week")) {
+                        days = nullSafeIntGet(valueToken) * 7;
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/interval/PGIntervalParser.java]: Why not use a `week` variable and populate the `week` field of `Period` ?

Comment (mkleen) [reply common/src/main/java/io/crate/interval/PGIntervalParser.java]: Weeks in psql are solely represented as days and never shown, therefore there is no need to handle them explicitly.

Comment (mkleen) [reply common/src/main/java/io/crate/interval/PGIntervalParser.java]:
```mkleen=# select interval '1 week';
 interval 
----------
 7 days
(1 row)
```
